applications/FemToDemApplication/python_scripts/CouplingFemDem.py

```diff
@@ -269,310 +234,312 @@ def Finalize(self):
 #============================================================================================================================
     def GenerateDEM(self): # This method creates the DEM elements and remove the damaged FEM, Additionally remove the isolated elements
 
-        FEM_elements = self.FEM_Solution.main_model_part.Elements
-
-        # Loop Over Elements to find the INACTIVE ones and generate the DEM only once
-        for Element in FEM_elements:
-            is_active     = True
-            DEM_Generated = Element.GetValue(KratosFemDem.DEM_GENERATED)
-
-            if Element.IsDefined(KratosMultiphysics.ACTIVE):
-                is_active = Element.Is(KratosMultiphysics.ACTIVE)
-
-                NumberOfDEM = 0         # Number of nodes with DEM Associated
-                for node in range(0, self.number_of_nodes_element): # Loop over nodes of the FE
-                    Node = Element.GetNodes()[node]
-                    if Node.GetValue(KratosFemDem.IS_DEM) == True:
-                        NumberOfDEM += 1
-
-                if is_active == False and DEM_Generated == False: # Let's generate the remaining DEM of this FE
-                    # print("elemento borrado: ", Element.Id)
-
-                    # --------------------- 1ST SCENARIO -----------------------------
-                    if NumberOfDEM == 0: # we must create 3 DEM
-
-                        dist01  = self.CalculateDistanceBetweenNodes(Element.GetNodes()[0], Element.GetNodes()[1])
-                        dist02  = self.CalculateDistanceBetweenNodes(Element.GetNodes()[0], Element.GetNodes()[2])
-                        dist12  = self.CalculateDistanceBetweenNodes(Element.GetNodes()[1], Element.GetNodes()[2])
-
-                        # look to the node 1 --------------
-                        Radius1 = self.GetMinimumValue(dist01, dist02) * 0.5
-                        Coordinates1 = self.GetNodeCoordinates(Element.GetNodes()[0])
-                        Id1 = Element.GetNodes()[0].Id
+        if self.FEM_Solution.main_model_part.ProcessInfo[KratosFemDem.GENERATE_DEM]:
+            FEM_elements = self.FEM_Solution.main_model_part.Elements
 
-                        self.ParticleCreatorDestructor.FEMDEM_CreateSphericParticle(Coordinates1, Radius1, Id1)
-                        Element.GetNodes()[0].SetValue(KratosFemDem.IS_DEM, True)        # Has an asociated DEM now
-                        Element.GetNodes()[0].SetValue(KratosMultiphysics.RADIUS, Radius1)
+            # Loop Over Elements to find the INACTIVE ones and generate the DEM only once
```

Comment:
I am not going to enter into details, but this looks like can be faster if done in C. Be careful with the OMP. Specially if you are going to use DEM, where the python part can be a performance killer

Reply:
YES! Currently I am reimplementing this DEM generation into a c++ process. I agree